File: impuls/runtimes/nodejs/runtime.js
#!/usr/bin/env node

/**
 * Impuls Function Runtime - Node.js
 * 
 * This is the runtime that executes inside the Firecracker VM.
 * It receives function invocations via HTTP and executes the handler.
 */

const http = require('http');
const vm = require('vm');
const fs = require('fs');
const path = require('path');

const PORT = process.env.RUNTIME_PORT || 8080;
const FUNCTION_DIR = process.env.FUNCTION_DIR || '/var/task';

// Function cache
let cachedHandler = null;
let cachedCode = null;

/**
 * Load and compile the function code
 */
function loadFunction(code, handler) {
    if (cachedCode === code && cachedHandler) {
        return cachedHandler;
    }

    // Create a sandbox environment
    const sandbox = {
        module: { exports: {} },
        exports: {},
        require: require,
        console: console,
        process: process,
        Buffer: Buffer,
        setTimeout: setTimeout,
        setInterval: setInterval,
        setImmediate: setImmediate,
        clearTimeout: clearTimeout,
        clearInterval: clearInterval,
        clearImmediate: clearImmediate,
        __dirname: FUNCTION_DIR,
        __filename: path.join(FUNCTION_DIR, 'function.js'),
    };
    sandbox.exports = sandbox.module.exports;

    // Compile and run the code
    try {
        const script = new vm.Script(code, {
            filename: 'function.js',
            displayErrors: true,
        });

        const context = vm.createContext(sandbox);
        script.runInContext(context);

        // Get the handler function
        const handlerParts = handler.split('.');
        const handlerName = handlerParts[handlerParts.length - 1];

        const handlerFn = sandbox.module.exports[handlerName] || sandbox.exports[handlerName];

        if (typeof handlerFn !== 'function') {
            throw new Error(`Handler '${handlerName}' is not a function`);
        }

        cachedCode = code;
        cachedHandler = handlerFn;

        return handlerFn;
    } catch (err) {
        throw new Error(`Failed to load function: ${err.message}`);
    }
}

/**
 * Execute the function handler
 */
async function executeHandler(handler, event, context) {
    return new Promise((resolve, reject) => {
        const timeout = setTimeout(() => {
            reject(new Error('Function execution timed out'));
        }, context.getRemainingTimeInMillis());

        try {
            // Check handler signature
            if (handler.length <= 2) {
                // Async handler (event, context) => Promise
                Promise.resolve(handler(event, context))
                    .then(result => {
                        clearTimeout(timeout);
                        resolve(result);
                    })
                    .catch(err => {
                        clearTimeout(timeout);
                        reject(err);
                    });
            } else {
                // Callback handler (event, context, callback) => void
                handler(event, context, (err, result) => {
                    clearTimeout(timeout);
                    if (err) {
                        reject(err);
                    } else {
                        resolve(result);
                    }
                });
            }
        } catch (err) {
            clearTimeout(timeout);
            reject(err);
        }
    });
}

/**
 * Create Lambda-like context object
 */
function createContext(functionName, timeoutMs, memoryMB) {
    const startTime = Date.now();
    const requestId = generateRequestId();

    return {
        functionName: functionName,
        functionVersion: '$LATEST',
        invokedFunctionArn: `arn:impuls:function:${functionName}`,
        memoryLimitInMB: memoryMB,
        awsRequestId: requestId,
        logGroupName: `/impuls/${functionName}`,
        logStreamName: `${new Date().toISOString().split('T')[0]}/${requestId}`,
        callbackWaitsForEmptyEventLoop: true,
        getRemainingTimeInMillis: () => {
            return Math.max(0, timeoutMs - (Date.now() - startTime));
        },
        done: (err, result) => {
            // Legacy callback
        },
        fail: (err) => {
            // Legacy callback
        },
        succeed: (result) => {
            // Legacy callback
        },
    };
}

/**
 * Generate a request ID
 */
function generateRequestId() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
        const r = (Math.random() * 16) | 0;
        const v = c === 'x' ? r : (r & 0x3) | 0x8;
        return v.toString(16);
    });
}

/**
 * Handle function invocation request
 */
async function handleInvoke(req, res) {
    let body = '';

    req.on('data', chunk => {
        body += chunk.toString();
    });

    req.on('end', async () => {
        const startTime = Date.now();
        let logs = [];

        // Capture console output
        const originalLog = console.log;
        const originalError = console.error;
        const originalWarn = console.warn;

        console.log = (...args) => {
            logs.push({ level: 'info', message: args.join(' ') });
            originalLog.apply(console, args);
        };
        console.error = (...args) => {
            logs.push({ level: 'error', message: args.join(' ') });
            originalError.apply(console, args);
        };
        console.warn = (...args) => {
            logs.push({ level: 'warn', message: args.join(' ') });
            originalWarn.apply(console, args);
        };

        try {
            const payload = JSON.parse(body);
            const { code, handler, event, env, timeout_ms, memory_mb, function_name } = payload;

            // Set environment variables
            if (env) {
                Object.entries(env).forEach(([key, value]) => {
                    process.env[key] = value;
                });
            }

            // Load the function
            const handlerFn = loadFunction(code, handler);

            // Create context
            const context = createContext(
                function_name || 'anonymous',
                timeout_ms || 30000,
                memory_mb || 128
            );

            // Execute the handler
            const result = await executeHandler(handlerFn, event, context);

            // Restore console
            console.log = originalLog;
            console.error = originalError;
            console.warn = originalWarn;

            // Send response
            const response = {
                statusCode: 200,
                body: result,
                duration_ms: Date.now() - startTime,
                logs: logs.map(l => `[${l.level.toUpperCase()}] ${l.message}`).join('\n'),
            };

            res.writeHead(200, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify(response));

        } catch (err) {
            // Restore console
            console.log = originalLog;
            console.error = originalError;
            console.warn = originalWarn;

            const response = {
                statusCode: 500,
                error: err.message,
                stack: err.stack,
                duration_ms: Date.now() - startTime,
                logs: logs.map(l => `[${l.level.toUpperCase()}] ${l.message}`).join('\n'),
            };

            res.writeHead(500, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify(response));
        }
    });
}

/**
 * Health check handler
 */
function handleHealth(req, res) {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({
        status: 'healthy',
        runtime: 'nodejs',
        version: process.version,
    }));
}

/**
 * Main HTTP server
 */
const server = http.createServer((req, res) => {
    if (req.method === 'POST' && req.url === '/invoke') {
        handleInvoke(req, res);
    } else if (req.method === 'GET' && req.url === '/health') {
        handleHealth(req, res);
    } else {
        res.writeHead(404, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ error: 'Not found' }));
    }
});

server.listen(PORT, '0.0.0.0', () => {
    console.log(`Impuls Node.js runtime listening on port ${PORT}`);
});

// Handle graceful shutdown
process.on('SIGTERM', () => {
    console.log('Received SIGTERM, shutting down...');
    server.close(() => {
        process.exit(0);
    });
});

process.on('SIGINT', () => {
    console.log('Received SIGINT, shutting down...');
    server.close(() => {
        process.exit(0);
    });
});

File: impuls/runtimes/python/runtime.py
#!/usr/bin/env python3
"""
Impuls Function Runtime - Python

This is the runtime that executes inside the Firecracker VM.
It receives function invocations via HTTP and executes the handler.
"""

import http.server
import json
import os
import sys
import traceback
import importlib.util
import time
from typing import Any, Callable, Dict, Optional

PORT = int(os.environ.get('RUNTIME_PORT', 8080))
FUNCTION_DIR = os.environ.get('FUNCTION_DIR', '/var/task')

# Function cache
cached_handler: Optional[Callable] = None
cached_code: Optional[str] = None


class LambdaContext:
    """Simplified AWS Lambda-like context object"""
    
    def __init__(self, function_name: str, memory_mb: int, timeout_sec: int):
        self.function_name = function_name
        self.function_version = '1'
        self.memory_limit_in_mb = memory_mb
        self._timeout_sec = timeout_sec
        self._start_time = time.time()
    
    def get_remaining_time_in_millis(self) -> int:
        elapsed = time.time() - self._start_time
        remaining = max(0, (self._timeout_sec - elapsed) * 1000)
        return int(remaining)


def load_function(code: str, handler: str) -> Callable:
    """Load and compile the function code"""
    global cached_handler, cached_code
    
    if cached_code == code and cached_handler is not None:
        return cached_handler
    
    # Parse handler (format: "module.function_name")
    handler_parts = handler.rsplit('.', 1)
    if len(handler_parts) != 2:
        raise ValueError(f"Invalid handler format: {handler} (expected 'module.function')")
    
    handler_function = handler_parts[1]
    
    # Create a temporary module from the code
    function_path = os.path.join(FUNCTION_DIR, 'function.py')
    
    # Write the code to a file
    os.makedirs(FUNCTION_DIR, exist_ok=True)
    with open(function_path, 'w') as f:
        f.write(code)
    
    # Load the module
    spec = importlib.util.spec_from_file_location('function', function_path)
    if spec is None or spec.loader is None:
        raise RuntimeError("Failed to load function module")
    
    module = importlib.util.module_from_spec(spec)
    sys.modules['function'] = module
    spec.loader.exec_module(module)
    
    # Get the handler function
    if not hasattr(module, handler_function):
        raise ValueError(f"Handler '{handler_function}' not found in module")
    
    handler_fn = getattr(module, handler_function)
    
    if not callable(handler_fn):
        raise ValueError(f"Handler '{handler_function}' is not callable")
    
    cached_code = code
    cached_handler = handler_fn
    
    return handler_fn


async def execute_handler_async(handler: Callable, event: Any, context: LambdaContext) -> Any:
    """Execute an async handler"""
    import asyncio
    
    if asyncio.iscoroutinefunction(handler):
        return await handler(event, context)
    else:
        return handler(event, context)


def execute_handler(handler: Callable, event: Any, context: LambdaContext) -> Any:
    """Execute the function handler"""
    import asyncio
    
    if asyncio.iscoroutinefunction(handler):
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)
        try:
            return loop.run_until_complete(execute_handler_async(handler, event, context))
        finally:
            loop.close()
    else:
        return handler(event, context)


class RuntimeHandler(http.server.BaseHTTPRequestHandler):
    """HTTP request handler for the runtime"""
    
    def log_message(self, format: str, *args) -> None:
        """Suppress default logging"""
        pass
    
    def do_GET(self):
        """Health check endpoint"""
        if self.path == '/health':
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'status': 'healthy', 'runtime': 'python'}).encode())
        else:
            self.send_response(404)
            self.end_headers()
    
    def do_POST(self):
        """Handle function invocation"""
        if self.path != '/invoke':
            self.send_response(404)
            self.end_headers()
            return
        
        try:
            # Read request body
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length)
            request = json.loads(body.decode('utf-8'))
            
            # Extract invocation data
            code = request.get('code', '')
            handler_name = request.get('handler', 'handler.handler')
            event = request.get('event', {})
            env = request.get('env', {})
            function_name = request.get('function_name', 'unknown')
            memory_mb = request.get('memory_mb', 128)
            timeout_sec = request.get('timeout_sec', 30)
            
            # Set environment variables
            for key, value in env.items():
                os.environ[key] = value
            
            # Load the function
            handler = load_function(code, handler_name)
            
            # Create context
            context = LambdaContext(function_name, memory_mb, timeout_sec)
            
            # Execute the handler
            result = execute_handler(handler, event, context)
            
            # Send response
            response = {
                'statusCode': 200,
                'body': result
            }
            
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps(response).encode())
            
        except Exception as e:
            error_response = {
                'statusCode': 500,
                'error': str(e),
                'stack': traceback.format_exc()
            }
            
            self.send_response(200)  # Still 200, error in body
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps(error_response).encode())


def main():
    """Start the runtime server"""
    server = http.server.HTTPServer(('0.0.0.0', PORT), RuntimeHandler)
    print(f"Python runtime listening on port {PORT}", file=sys.stderr)
    
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        server.server_close()


if __name__ == '__main__':
    main()
